### ISSUE TITLE
Disconnect of intercommunicators is failing

### ISSUE THREAD

**Issue by rhc54** [open-mpi/ompi#1779]
When someone calls MPI_Comm_free or MPI_Comm_disconnect on an intercommunicator, we enter a function called ```ompi_dpm_base_disconnect_init```. That function attempts to execute the following code:

```c
    /* initiate all isend_irecvs. We use a dummy buffer stored on
       the object, since we are sending zero size messages anyway. */
    for ( i=0; i < obj->size; i++ ) {
        ret = MCA_PML_CALL(irecv (&(obj->buf), 0, MPI_INT, i,
                     OMPI_COMM_BARRIER_TAG, comm,
                     &(obj->reqs[2*i])));

        if ( OMPI_SUCCESS != ret ) {
            opal_output(0, "%s dpm_base_disconnect_init: error %d in irecv to process %d\n",
                        OMPI_NAME_PRINT(OMPI_PROC_MY_NAME), ret, i);
            free (obj->reqs);
            free (obj);
            return NULL;
        }
        ret = MCA_PML_CALL(isend (&(obj->buf), 0, MPI_INT, i,
                     OMPI_COMM_BARRIER_TAG,
                     MCA_PML_BASE_SEND_SYNCHRONOUS,
                     comm, &(obj->reqs[2*i+1])));

        if ( OMPI_SUCCESS != ret ) {
            opal_output(0, "%s dpm_base_disconnect_init: error %d in isend to process %d\n",
                        OMPI_NAME_PRINT(OMPI_PROC_MY_NAME), ret, i);
            free (obj->reqs);
            free (obj);
            return NULL;
        }
    }
```

On the parent process, this code fails every time with an error on the isend side:

```shell
[bend001:30668] [[64151,1],0] dpm_base_disconnect_init: error -12 in isend to process 0
```
When we do a loop_spawn, these messages queue up until the process consumes all available resources, at which point the spawning loop fails. If you are using TCP, then it will loop for a long time as there are lots of buffers available, and you get a deluge of the above error messages when the parent process exits.

It would appear to first inspection that process 0 is in fact the parent process, in which case this makes no sense - we don't need to send a process to ourselves, though it begs the question as to why that isend would fail. Can someone crosscheck this function?

**Comment by edgargabriel** [comment]
its been a while that I looked at this code, but keep in mind that this is a inter-communicator. Sending a message to process 0 means its process zero from the other group, not to yourself. This function basically mimics a non-blocking barrier operation. The code looks generally ok to me, the only question that I would is why do we use SEND_SYNCHRONOUS instead of the standard send, but even in that case that should not raise an error..

**Comment by rhc54** [comment]
Is it possible the remote process is exiting before the send can be completed? I'm trying to figure out how it is that the remote proc is "unreachable"

**Comment by edgargabriel** [comment]
this should not be possible if both groups call disconnect. However, it could happen if one of the two groups does not call disconnect.

**Comment by rhc54** [comment]
hmmm...the child code is certainly calling disconnect, but i wonder if it doesn't realize that it's an intercomm on its side? something is clearly borked.

**Comment by ggouaillardet** [comment]
If comm is an intra communicator, then we do isend/irecv with ourself.
My first impression is that this is dumb but harmless.

**Comment by ggouaillardet** [comment]
Are you referring the tests in orte/tests from v1.10 ? or ompi-test repo ?
The loop_spawn test from the ompi-tests repo is a bit different
In ompi-tests, both loop-spawn and loop_child call
MPI_Comm_disconnect(intercomm)
No such thing in v1.10 ...

I am pretty sure MPI_Comm_disconnect is somehow called in MPI_Finalize()
That makes no difference in loop_child, but makes a big one in loop_spawn.
(All 1000 tasks are released at the end, a memory exhaustion might occur before)

@bosilca can you please comment ?
Is the test in v1.10 correct ?
If yes, do we hit a memory exhaustion or a bug ?

**Comment by rhc54** [comment]
In orte/tests they call comm_free, which calls disconnect under the covers. However, I tried changing it to directly call disconnect without getting any difference in behavior.

**Comment by ggouaillardet** [comment]
@rhc54 which branch ?
Master is ok, v1.10 is not
You need to 
MPI_Comm_free(merged_communicator);
__and__
MPI_Comm_disconnect(intercomm);

I am afk and will double check that tomorrow only

**Comment by rhc54** [comment]
Ah, good point - I updated the orte/test version in 1.10 and it now runs cleanly. So it sounds like this is actually an issue with the tests instead of the code.

**Comment by rhc54** [comment]
I agree that the orte/test in master and 2.x is okay, and I checked ompi-tests and that looks okay too. So I think things look okay, and the problem may be in the user's code if they followed the orte/test examples. I fixed the 1.10 example.
